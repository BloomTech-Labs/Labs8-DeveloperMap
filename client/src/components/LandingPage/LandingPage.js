import React from 'react';
import axios from 'axios';
import 'mapbox-gl/dist/mapbox-gl.css';
import MapGL, { Marker, Popup } from 'react-map-gl';
import Geocoder from 'react-map-gl-geocoder';
import { NavLink } from 'react-router-dom';

import {
  MapWindow,
  ShowMarker,
  CloseX,
  PopupInfo,
  LogoImg,
  PinKey,
  KeyBox,
  ToggleKnob,
} from './MapWindowStyle';
import SeekerPin from '../../images/markerlogo.png';
import MainLogo from '../../images/mainlogo.png';
import CompanyPin from '../../images/markerlogo4.png';

const MAPBOX_TOKEN =
  'pk.eyJ1IjoibG5kdWJvc2UiLCJhIjoiY2pvNmF1ZnowMGo3MDNrbmw4ZTVmb2txMyJ9.UpxjYyEOBnCJjw_qE_N8Kw';

class LandingPage extends React.Component {
  state = {
    viewport: {
      latitude: 37.7577,
      longitude: -100,
      zoom: 3,
      width: '100%',
      height: '100%',
    },
    data: [],
    pin: null,
    filter: {
      seeker: true,
      company: true,
    },
  };

  mapRef = React.createRef();

  componentDidMount() {
    window.addEventListener('resize', this.resize);
    this.resize();
    this.getMarkers();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.resize);
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.currentSignedInUser !== this.props.currentSignedInUser &&
      this.props.currentSignedInUser !== null
    ) {
      if (this.props.currentSignedInUser.role === 'seeker') {
        this.setState({
          currentSignedInUser: this.props.currentSignedInUser,
          filter: {
            seeker: false,
            company: true,
          },
        });
      } else if (this.props.currentSignedInUser.role === 'company') {
        this.setState({
          currentSignedInUser: this.props.currentSignedInUser,
          filter: {
            seeker: true,
            company: false,
          },
        });
      }
    } else if (
      prevProps.currentSignedInUser !== this.props.currentSignedInUser &&
      this.props.currentSignedInUser === null
    ) {
      this.setState({
        filter: {
          seeker: true,
          company: true,
        },
      });
    }

    if (prevProps.location.pathname !== this.props.location.pathname) {
      this.getMarkers();
    }
  }

  resize = () => {
    this.handleViewportChange({
      width: window.innerWidth,
      height: window.innerHeight,
    });
  };

  handleViewportChange = viewport => {
    this.setState({
      viewport: { ...this.state.viewport, ...viewport },
    });
  };

  // ==== Gets the markers from the server ====
  getMarkers = () => {
    axios
      .get('https://intense-stream-29923.herokuapp.com/api/markers')
      .then(response => {
        const markerArray = [];
        for (let mark in response.data) {
          markerArray.push(response.data[mark]);
        }
        this.setState({ data: markerArray });
      })
      .catch(err => console.log(err));
  };

  // ==== Filter feature for the markers ====
  markerShow = e => {
    if (e.target.name === 'seeker') {
      this.setState(prevState => {
        return {
          filter: {
            ...this.state.filter,
            seeker: !prevState.filter.seeker,
          },
        };
      });
    } else if (e.target.name === 'company') {
      this.setState(prevState => {
        return {
          filter: {
            ...this.state.filter,
            company: !prevState.filter.company,
          },
        };
      });
    }
  };

  // ==== Renders the markers to the map ====
  renderMarker = (mark, i) => {
    if (mark.properties.role === 'seeker') {
      return (
        <Marker
          key={i}
          latitude={mark.geometry.coordinates[1]}
          longitude={mark.geometry.coordinates[0]}
          offsetTop={-40}
          offsetLeft={-25}
        >
          <ShowMarker
            src={SeekerPin}
            alt="red marker"
            onClick={() => this.setState({ pin: mark })}
            show={this.state.filter.seeker}
          />
        </Marker>
      );
    } else if (mark.properties.role === 'company') {
      return (
        <Marker
          key={i}
          latitude={mark.geometry.coordinates[1]}
          longitude={mark.geometry.coordinates[0]}
          offsetTop={-40}
          offsetLeft={-25}
        >
          <ShowMarker
            src={CompanyPin}
            alt="gold marker"
            onClick={() => this.setState({ pin: mark })}
            show={this.state.filter.company}
          />
        </Marker>
      );
    }
  };

  // ==== Renders the popup to the map ====
  renderPopup = () => {
    const { pin } = this.state;

    let fullName = '';

    if (!!pin) {
      if (!!pin.properties.title.firstName) {
        fullName = `${pin.properties.title.firstName} ${
          pin.properties.title.lastName
        }`;
        return (
          pin && (
            <Popup
              latitude={pin.geometry.coordinates[1]}
              longitude={pin.geometry.coordinates[0]}
              offsetTop={-20}
              closeButton={false}
              closeOnClick={false}
            >
              <CloseX onClick={() => this.setState({ pin: null })}>X</CloseX>
              <PopupInfo
                style={{ cursor: 'pointer' }}
                onClick={() =>
                  this.props.history.push(`/seeker/${pin.properties.uid}`)
                }
              >
                {fullName}
              </PopupInfo>
            </Popup>
          )
        );
      } else if (!!pin.properties.title.companyName) {
        fullName = pin.properties.title.companyName;
        return (
          pin && (
            <Popup
              latitude={pin.geometry.coordinates[1]}
              longitude={pin.geometry.coordinates[0]}
              offsetTop={-20}
              closeButton={false}
              closeOnClick={false}
            >
              <CloseX onClick={() => this.setState({ pin: null })}>X</CloseX>
              <PopupInfo
                style={{ cursor: 'pointer' }}
                onClick={() =>
                  this.props.history.push(`/employer/${pin.properties.uid}`)
                }
              >
                {fullName}
              </PopupInfo>
            </Popup>
          )
        );
      }
    }
  };

  render() {
    return (
      <MapWindow>
        <MapGL
          ref={this.mapRef}
          mapboxApiAccessToken={MAPBOX_TOKEN}
          {...this.state.viewport}
          onViewportChange={this.handleViewportChange}
          style={{ position: 'absolute', margin: '-1px' }}
          mapStyle="mapbox://styles/lndubose/cjohrsfn608in2qqyyn2wu15g"
          onClick={() => this.setState({ pin: null })}
        >
          <Geocoder
            mapRef={this.mapRef}
            mapboxApiAccessToken={MAPBOX_TOKEN}
            onViewportChange={this.handleViewportChange}
            placeholder="Search for places"
          />
          <LogoImg alt="logo" src={MainLogo} />
          <KeyBox>
            <div className="key">
              <PinKey src={SeekerPin} />
              <h3>Job Seeker</h3>
              <ToggleKnob htmlFor="seeker">
                <input
                  type="checkbox"
                  name="seeker"
                  id="seeker"
                  checked={this.state.filter.seeker}
                  onChange={this.markerShow}
                />
              </ToggleKnob>
            </div>
            <div className="key">
              <PinKey src={CompanyPin} />
              <h3>Employer</h3>
              <ToggleKnob htmlFor="company">
                <input
                  type="checkbox"
                  name="company"
                  id="company"
                  checked={this.state.filter.company}
                  onChange={this.markerShow}
                />
              </ToggleKnob>
            </div>
            <NavLink to='/tutorial'>How it Works</NavLink>
          </KeyBox>

          {this.state.data.map(this.renderMarker)}

          {this.renderPopup()}
        </MapGL>
      </MapWindow>
    );
  }
}

export default LandingPage;
