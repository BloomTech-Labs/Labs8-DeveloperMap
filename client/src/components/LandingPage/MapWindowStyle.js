import styled from 'styled-components';

export const MapWindow = styled.div`
  .mapboxgl-ctrl-geocoder {
    width: 100%;
    border-radius: 20px;

    @media (max-width: 1050px) {
      min-width: 200px;
    }

    @media (max-width: 350px) {
      min-width: 170px;
    }
    .geocoder-icon-search {
      top: 7px;
      left: 3%;
    }

    input[type='text'] {
      height: auto;
      width: 100%;
      border-radius: 20px;
      border: none;
      padding: 10px 0 10px 14%;
      &:focus {
        outline: none;
      }
    }
    .suggestions {
      background-color: white;
      border-radius: 4px;
      padding-inline-start: 20px;
      li {
        list-style-type: none;
        margin: 2px 0;
        cursor: pointer;
        &:hover {
          color: rgb(54, 154, 255);
        }
      }
      .active {
        color: rgb(119, 187, 255);
      }
    }
  }

  .mapboxgl-control-container {
    width: 100%;
    .mapboxgl-ctrl-top-right {
      margin-left: Calc(3% + 10px);
      margin-top: 5px;
      right: initial;
      top: initial;
      left: initial;

      @media (max-width: 1050px) {
        margin-left: Calc(3% + 10px);
      }

      .mapboxgl-ctrl {
        margin: 10px 0 0 0;
      }
    }
  }
  .geocoder-pin-right {
    display: none;
  }

  .mapboxgl-popup {
    height: 50px;
  }

  .mapboxgl-popup-content {
    padding: 0;
    border-radius: 20px;
  }

  .mapboxgl-popup-anchor-bottom {
    .mapboxgl-popup-tip {
      border-top-color: rgb(35, 34, 48);
    }
  }
`;

export const ShowMarker = styled.img`
  height: 35px;
  width: 25px;
  cursor: pointer;
  

  &.company {
    display: ${props => (props.show.company ? 'block' : 'none')};
  }

  &.seeker {
    display: ${props => (props.show.seeker ? 'block' : 'none')};
  }
`;

export const CloseX = styled.p`
  position: absolute;
  right: -7px;
  top: -3px;
  font-size: 19px;
  cursor: pointer;
  background-color: rgb(37, 36, 48);
  color: rgb(255,255,255);
  border-radius: 50%;
  height: 20px;
  width: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const PopupInfo = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  max-width: 210px;
  padding: 8px 0px 0px 0px;

  .container {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 8px 6px;

  }

  div {
    text-align: center;

    h4 {
      margin-bottom: 7px;
    }

    p {
      font-size: 0.9rem;
    }

    .jobTitle {
      display: flex;
      flex-wrap: wrap;
      justify-content: center;

      p {
        padding-left: 2px;
      }
    }
  }
`;

export const Link = styled.div`
  color: white;
  cursor: pointer;
  width: 100%;
  background: rgb(35, 34, 48);
  border-bottom-right-radius: 20px;
  border-bottom-left-radius: 20px;
  padding: 5px 0px;
`;

export const PopupImg = styled.div`
  display: ${props => (props.image ? 'block' : 'none')};
  width: 60px;
  height: 60px;
  background-image: ${props => (props.image ? `url(${props.image})` : 'null')};
  background-size: cover;
  background-repeat: no-repeat;
  border-radius: 50%;
  background-position: top;
  margin-right: 5px;
`;

export const LogoImg = styled.img`
  position: absolute;
  bottom: 0;
  left: 0;
  width: 300px;
  height: auto;
  margin: 2%;
  @media (max-width: 550px) {
    margin: 50px auto;
    padding-left: 30%;
    width: 250px;
  }
`;

export const PinKey = styled.img`
  width: 30px;
  height: 45px;
`;

export const KeyBox = styled.div`
  width: 205px;
  height: 135px;
  background-color: rgba(109, 7, 26, 0.95);
  position: absolute;
  right: 0;
  bottom: 0;
  margin: 4% 3%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border-radius: 20px;
  z-index: 5;

  .container {
    padding: 6px 3%;
    background-color: #fafafa;
    border-radius: 10px;

    @media (max-width: 550px) {
      display: flex;
      padding: 0 2%;
    }
  }

  .key {
    display: flex;
    align-items: center;
    padding: 2px 0;
  }
  h3 {
    font-size: 1.1rem;
    font-weight: lighter;
    width: 90px;
  }

  @media (max-width: 550px) {
    bottom: 0;
    left: 0;
    margin: 0;
    width: 100%;
    flex-direction: row;
    height: 50px;
    border-radius: 0;
  }
`;

export const ToggleKnob = styled.label`
  position: relative;
  width: 30px;
  height: 12px;
  background-color: ${props =>
    props.children.props.checked
      ? props.children.props.id === 'company'
        ? 'rgb(207, 149, 4)'
        : 'rgb(122, 38, 38)'
      : 'rgb(128, 128, 128)'};
  margin: 20px 0;
  border-radius: 5px;
  z-index: 10;
  margin: 0 10px;
  cursor: pointer;

  input[type='checkbox'] {
    outline: none;
    appearance: none;
    display: block;
    width: 17px;
    height: 17px;
    border-radius: 50%;

    transition: all 0.5s ease;
    cursor: pointer;
    position: absolute;
    top: -5px;
    left: 15px;

    background: rgb(37, 36, 48);
  }

  input[type='checkbox']:checked {
    left: -5px;
  }
`;


export const PopContent = styled.div`
  width: 128px;
`;