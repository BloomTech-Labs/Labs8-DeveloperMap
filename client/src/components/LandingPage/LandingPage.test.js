import React from 'react';
import renderer from 'react-test-renderer';

import LandingPage from './LandingPage';

describe(`Landing Page Component`, () => {
    it('renders correctly', () => {
        const tree = renderer.create(<LandingPage />).toJSON();
        expect(tree).toMatchSnapshot();
      });
})