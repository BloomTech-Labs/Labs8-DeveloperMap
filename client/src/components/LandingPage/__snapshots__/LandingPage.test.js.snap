// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`landing page component should render without error 1`] = `
<div
  className="sc-bdVaJa sIavw"
>
  <div
    style={
      Object {
        "cursor": "grab",
        "height": 768,
        "margin": "-1px",
        "position": "absolute",
        "width": 1024,
      }
    }
  >
    <div
      style={
        Object {
          "height": "100%",
          "position": "relative",
          "width": "100%",
        }
      }
    >
      <div
        className=""
        style={
          Object {
            "height": "100%",
            "overflow": "hidden",
            "position": "absolute",
            "visibility": "inherit",
            "width": "100%",
          }
        }
      />
      <div
        style={
          Object {
            "height": 0,
            "overflow": "visible",
            "width": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;
